Remove pos

mathjeff opened the pull request.

mathjeff committed:
adding compare.go to facilitate smarter parser_test diffing

this should help when making more test cases

Bug: b/34671474
Test: make blueprint_tools

Change-Id: Iff6197ac64053fde33119bbf59885c714b27c91f

mathjeff commented:
androidmk patch is https://android-review.googlesource.com/c/333894/

colincross commented on parser/parser_test.go:
```diff
@@ -639,23 +639,276 @@ func TestParseValidInput(t *testing.T) {
+	if actual.item.Kind() != expected.item.Kind() {
+		return false, fmt.Sprintf("%s is of type %s whereas %s is of type %s", actual.name, actual.item.Kind(), expected.name, expected.item.Kind())
+	}
+	switch actual.item.Kind() {
```
This conversion is unnecessary, use
`reflect.DeepEqual(actual.item.Interface(), expected.item.Interface())`

mathjeff replied:
cool

mathjeff replied:
Well I investigated a little further but it seems that unfortunately, x.Interface() fails if there were any private fields traversed while accessing x. My thought now is that it's preferable to have support for diffs of private fields if it requires reimplementing this method like this.

colincross commented on parser/parser_test.go:
```diff
@@ -639,23 +639,276 @@ func TestParseValidInput(t *testing.T) {
+	aKeys := a.MapKeys()
+	bKeys := b.MapKeys()
+	for _, aKey := range aKeys {
+		if !(isIn(aKey, bKeys)) {
```
This is unnecessary, use:
if !b.MapIndex(aKey).IsValid()

mathjeff replied:
See below about keys being pointers

colincross replied:
What happens if two deeply equal pointers are keys in the same map?  Only the pointer value will be hashed, so they will still be two different keys.

mathjeff replied:
Yeah exactly; the test doesn't confirm that the pointers themselves are correct, only that they have a corresponding value in the other map.

Do you mean that we want to fix deepCompare for the following case?

////
var a and b, deeply equal to each other
var c and d, deeply equal to each other

map1{a:true, b:true, c:true}
map2{a:true, c:true, d:true}

?

for which deepCompare would return true when really we want false

colincross replied:
We would need to define what the semantics are for comparing two maps whose keys are deepequal but not equal.  Even for two maps that were deep-cloned off of each other, there is no guarantee that we can determine a 1-1 mapping between keys in the two maps.

The alternative is to sidestep this problem and refuse to deepcompare pointers or anything that might contain pointers (structs, arrays) as the keys in maps.  This should be fine for the parser/printer, since they only support strings as the keys to maps.

mathjeff replied:
Unfortunately the test needs to be able to compare SyntaxTree.comments, whose type is map[ParseNode](*CommentPair) , which uses an interface type as its key

I think that my inclination, if we want to spend the effort, would be two changes. #1 we fix the case in deepCompare involving maps having recursively equal keys such that deepCompare deems them equal, and #2 we require the individual tests to explicitly invoke deepCompare, so if the tests require certain relationships between the pointers themselves (currently the checks specific to the pointers are in the Builder class itself), then the tests can validate those requirements separately, and can still use deepCompare while doing it.

The changes to deepCompare I imagine would be to have it assume that if there exists a 1-1 mapping for which we cannot find a relevant difference, then we deem the data structures equivalent. In other words: 1. For each combination of entries in map1 and map2, check whether those two entries are equal. 2. For each entry A in map1 that has K matches in map2, confirm that there are exactly K such entries in map1 that have the same set of matches in map2 3. If every condition in #2 yielded true, then return true.


As for sidestepping the problem, my favorite option is to invoke the printer (or perhaps VerbosePrinter) in printer_test.go and check the text output.


Alternatively we could adjust the code being tested attach an integer id to each ParseNode and to use that as the map key rather than the ParseNode reference itself.

colincross commented on parser/parser_test.go:
```diff
@@ -639,23 +639,276 @@ func TestParseValidInput(t *testing.T) {
+	return item, false
+}
+
+func isIn(item reflect.Value, list []reflect.Value) (contained bool) {
```
I think this function (and findMatching and compareValues) are unnecessary due to comments below

mathjeff replied:
See below about keys being pointers

colincross commented on parser/parser_test.go:
```diff
@@ -639,23 +639,276 @@ func TestParseValidInput(t *testing.T) {
+		}
+	}
+	for _, aKey := range aKeys {
+		bKey, _ := findMatching(aKey, bKeys)
```
This seems unnecessary, findMatching is going to return a key that is deeply equal to aKey, so why not just use aKey?

mathjeff replied:
Yeah, I was doing a direct map lookup initially. Since the keys are themselves pointers pointing to different but recursively equal objects, the lookup wasn't finding the match.

Technically there's a detail here that doesn't get tested directly by the test now, which is that the key in the map not only should be recursively equal to another map key in the original object, but also that it should be the same reference as another object in the parse tree somewhere. For the moment that check is in the builder class itself though.

The more that I work on this parser test, the more I wonder if it would be better to have the parser test use the printer and do a text diff instead, and then to reimplement the printer test to directly pass parseTrees into the printer and to not have it use the parser at all. I guess my default is to not bother to rearrange the parser and printer tests, but I could do that too if you think it would be better. Technically that approach wouldn't be able to rule out the possibility that the parser was outputting some unexpected data structure that the printer just so happened to accept and just so happened to print in a way that matched the desired output - I hypothesize we wouldn't be very concerned about that possibility either.

colincross commented on parser/parser_test.go:
```diff
@@ -639,23 +639,276 @@ func TestParseValidInput(t *testing.T) {
+	b := expected.item
+	aLen := a.Len()
+	bLen := b.Len()
+	sharedLen := int(math.Min(float64(aLen), float64(bLen)))
```
Don't cast to float, either implement a trivial min function or open code it

mathjeff replied:
ok

colincross commented on parser/parser_test.go:
```diff
@@ -639,23 +639,276 @@ func TestParseValidInput(t *testing.T) {
+	aLen := a.Len()
+	bLen := b.Len()
+	sharedLen := int(math.Min(float64(aLen), float64(bLen)))
+	for i := 0; i < sharedLen; i++ {
```
Or: 
`for i := 0; i < a.Len() && i < b.Len(); i++ {`

mathjeff replied:
I'll re-implement the 'min' function as you mentioned above.

I extracted the 'min' operation to slightly decrease the number of values being recomputed during each loop iteration (I don't see anything in https://github.com/golang/go/wiki/CompilerOptimizations specifying that the Go compiler can optimize 'i < a.Len() && i < b.Len()').

colincross commented on parser/parser_test.go:
```diff
@@ -639,23 +639,276 @@ func TestParseValidInput(t *testing.T) {
+	b := expected.item
+	aCount := a.NumField()
+	bCount := b.NumField()
+	if aCount != bCount {
```
It would be more reliable to compare the types here, or delete the check entirely

mathjeff replied:
ok

colincross commented on parser/parser_test.go:
```diff
@@ -639,23 +639,276 @@ func TestParseValidInput(t *testing.T) {
+	return true, ""
+}
+
+// deepCompare is much like reflect.DeepEqual, but it does a deep comparison of map keys too (which is slower than a map lookup)
```
Deep comparing map keys seems unnecessary to me.  If they hash to the same value they must be == equal, and if they are == equal then they must be DeepEqual.  If that is not true, can you construct a test that returns different values for reflect.DeepEqual and deepCompare?

mathjeff replied:
Yeah, I was surprised too. I'll add a test involving a map having keys of pointer type.

colincross commented on parser/parser_test.go:
```diff
@@ -639,23 +639,276 @@ func TestParseValidInput(t *testing.T) {
+	return deepCompareObjects(namedItem{"actual", reflect.ValueOf(actual)}, namedItem{"expected", reflect.ValueOf(expected)}, 0)
+}
+
+// deprecated
```
Why check in a deprecated function?

mathjeff replied:
I'll remove it

colincross commented on parser/parser.go:
```diff
@@ -37,77 +37,193 @@ func (e *ParseError) Error() string {
-		Comments: comments,
-	}, errs
-
+type parser struct {
```
reordering these makes this really hard to review

mathjeff replied:
done

colincross commented on parser/parser.go:
```diff
@@ -37,77 +37,193 @@ func (e *ParseError) Error() string {
-	Name     string
-	Defs     []Definition
-	Comments []*CommentGroup
+type ParseTree struct {
```
The distinction between ParseTree and SyntaxTree is not clear to me.  Leaving this as File and File.Name would be much easier to review, and then you can use a separate patch to do renames.  The gorename command line tool is useful for doing programmatic renames, and you can include the gorename command in the commit message.

mathjeff replied:
done

mathjeff commented on parser/parser_test.go:
```diff
@@ -639,23 +639,276 @@ func TestParseValidInput(t *testing.T) {
 			t.Errorf("test case: %s", testCase.input)
 			t.Errorf("length mismatch, expected %d definitions, got %d",
 				len(testCase.defs), len(file.Defs))
+
```
good idea

mathjeff commented on parser/parser_test.go:
```diff
@@ -639,23 +639,276 @@ func TestParseValidInput(t *testing.T) {
+	return found
+}
+
+// checks if two objects are the same class, returns True if they match
```
ok

mathjeff committed:
Preemptively rename File to ParseTree to clarify its upcoming meaning.

A ParseTree represents a tree that was parsed from text, and also
includes information about where its nodes came from. Currently the
source information is attached to the nodes themselves, but that's
changing in a subsequent patch, at which point the ParseTree will
directly contain the source information. At that point, it will be
possible for a programmatic builder class to generate a syntax tree
without having to think about or pass in empty values for text
positions.

Really the reason that this patch is separate is to make the code review
more convenient.

gorename -from 'github.com/google/blueprint/parser.File.Name' -to 'FileName'
gorename -from 'github.com/google/blueprint/parser.File' -to 'ParseTree'

Bug: b/34671474
Test: make blueprint_tools

Change-Id: I7d591ac123ccd1697a3d739308596655afc39197

mathjeff committed:
Update Go version in Travis CI, to make consistent the output observed by compare_test.go

Bug: b/34671474
Test: upload a patch for Travis to build

Change-Id: I1bb7f94104e904c09c4175c5c85164080c6756fc

colincross commented on parser/printer.go:
```diff
@@ -330,36 +330,60 @@ func (p *printer) flush() {
-func (p *printer) unindent(pos scanner.Position) {
-	p.printEndOfLineCommentsBefore(pos)
-	p.indentList = p.indentList[0 : len(p.indentList)-1]
+func (p *VerbosePrinter) Print() string {
```
This is the only public method, and always used as NewVerbosePrinter(...).Print(), might as well make a VerbosePrint(...) method that delegates to an unexported verbosePrinter type.

mathjeff replied:
ok

colincross commented on parser/printer_test.go:
```diff
@@ -315,34 +315,67 @@ test { // test
+		if in[0] == '\n' {
+			in = in[1:]
+		} else {
+			panic(fmt.Sprintf("A newline at the beginning of testcase input is required, to make the test code easily readable. Offending test case '%s' starts with character: '%s'", description, string(in[0])))
```
This doesn't seem worth enforcing, just strip it if it is there.

mathjeff replied:
ok

colincross commented on proptools/compare.go:
```diff
@@ -0,0 +1,248 @@
+	Item interface{}
+}
+
+type Comparison struct {
```
a two-value return is more idiomatic than returning a struct

mathjeff replied:
When is that (two values preferable to a struct) the case? Is it when the two values have fundamentally different meanings? It seems to me that a Comparison is itself a concept. If we wanted to instead return an entire list of differences, it'd be convenient to not have to update the signatures of each method based on the implementation of a Comparison.

colincross replied:
I'd say that a struct should be used if you are going to pass that encapsulated value around (more than just returning it).  If you are going to immediately unpack and operate on it, a multiple-value return is better.

mathjeff replied:
That logic sounds like a decent rule of thumb to me: to use a multi-valued return if the caller is likely to immediately read all of the contained fields. Does it apply here? None of the functions in compare.go read the field Comparison.Difference .

If for example we wanted to add another field to Comparison that perhaps specified the property path of the difference, then those fields probably wouldn't be read by anything in compare.go either.

mathjeff replied:
I talked to Dan and he thinks that if the number of fields is small then the thing to do is to make the fields be separate return values, and if the number of fields grows then to consider making the return be a struct at that time. I suppose that that's enough of a rule of thumb for me to be able to follow anyway, so I'll do that.

colincross commented on proptools/compare.go:
```diff
@@ -0,0 +1,248 @@
+	}
+}
+
+// compares two maps
```
document the semantics of this comparison, especially as compared to reflect.DeepEqual

mathjeff replied:
done

colincross commented on proptools/compare.go:
```diff
@@ -0,0 +1,248 @@
+	return equal
+}
+
+// compares two arrays
```
compares two arrays or slices

document that this compares [0,len()), and not [0,cap())

mathjeff replied:
ok

colincross commented on proptools/compare.go:
```diff
@@ -0,0 +1,248 @@
+	}
+}
+
+// compares two pointers in the same manner as reflect.deepEqual
```
compares two pointers or interfaces

mathjeff replied:
done

colincross reviewed (commented):
Partial review of the first 2 commits

mathjeff committed:
Some cleanup of the parser by removing Pos from ParseNode

Now the final position of each node is now decided separately by the
printer during printing, based on the node's place in the SyntaxTree

The parser now generates a ParseTree, which has slightly more
information than a SyntaxTree - it contains both the syntax tree and the
original source positions of the tokens from which the syntax tree was
parsed

Bug: b/34671474
Test: make blueprint_tools

Change-Id: I55466114917cc88ac9e4e60b56bda2d4318c231a

mathjeff commented on parser/parser_test.go:
```diff
@@ -16,7 +16,7 @@ package parser
 import (
 	"bytes"
-	"reflect"
+	"github.com/google/blueprint/proptools"
```
Have you made any customizations to Gogland that automate things like line lengths and import ordering? If so, I wonder if we can check them in somewhere.

colincross replied:
No, and it keeps screwing them up for me too.

mathjeff commented on parser/parser_test.go:
```diff
@@ -639,22 +639,123 @@ func TestParseValidInput(t *testing.T) {
+	expectedNodes := expected.ListOfAllNodes()
+	remappings = make(map[ParseNode]ParseNode, 0)
+	for i, actual := range actualNodes {
+		remappings[actual] = expectedNodes[i]
```
ok

mathjeff commented on parser/parser_test.go:
```diff
@@ -639,22 +639,123 @@ func TestParseValidInput(t *testing.T) {
+
+	// now compare the comments
+	return proptools.DeepCompare(
+		proptools.NamedItem{"expected.SyntaxTree.Comments", expected.SyntaxTree.comments},
```
ok

mathjeff commented on parser/parser_test.go:
```diff
@@ -639,22 +639,123 @@ func TestParseValidInput(t *testing.T) {
+		proptools.NamedItem{"actual.SyntaxTree.Comments", analogousActualComments})
+}
+
+func compareSourcePositions(expected *ParseTree, actual *ParseTree, actualNode_remappings map[ParseNode]ParseNode) proptools.Comparison {
```
Yeah I don't like the underscore either; I just use it to clarify that this is the remappings of the actualNodes as opposed to being the actual nodeRemappings. What name would you recommend that conveys as much information?

colincross replied:
Sometimes its better to use a short, readable name and put the extra information in a comment, rather than carrying all the information everywhere you use the name.  See https://github.com/golang/go/wiki/CodeReviewComments#variable-names

mathjeff replied:
Hmm it seems to me that the trouble in this case is more that I forgot to add a comment than that I used a long variable name.

In any case, I've now redone it in the next version of this commit for you to take a look at.

mathjeff commented on parser/parser_test.go:
```diff
@@ -639,22 +639,123 @@ func TestParseValidInput(t *testing.T) {
+	// The reason we replace the keys is so proptools.DeepCompare can know which keys we want to compare against which others
+
+	analogousSourcePositions := make(map[ParseNode]scanner.Position)
+	for actualKey, actualValue := range expected.SourcePositions {
```
true

mathjeff commented on parser/parser_test.go:
```diff
@@ -639,22 +639,123 @@ func TestParseValidInput(t *testing.T) {
+
+	// use the node mapping to compare the comments
+	commentsComparison := compareComments(expected, actual, remappings)
+	if !commentsComparison.Equal {
```
ok

mathjeff commented on parser/parser_test.go:
```diff
@@ -639,22 +639,123 @@ func TestParseValidInput(t *testing.T) {
+	//return compareSourcePositions(expected, actual, remappings)
+}
+
+func runValidIndex(t *testing.T, i int) {
```
ok

mathjeff commented on parser/parser_test.go:
```diff
@@ -639,22 +639,123 @@ func TestParseValidInput(t *testing.T) {
+	expectedRepresentation := NewVerbosePrinter(correctParse.SyntaxTree).Print()
+	actualRepresentation := NewVerbosePrinter(actualParse.SyntaxTree).Print()
+	if !comparison.Equal {
+		t.Errorf(`
```
ok

mathjeff commented on parser/parser_test.go:
```diff
@@ -639,22 +639,123 @@ func TestParseValidInput(t *testing.T) {
+
+`, i, testCase.input, expectedRepresentation, actualRepresentation, comparison.Difference)
+	}
+	succeeded = true
```
thanks

mathjeff commented on parser/printer.go:
```diff
@@ -330,36 +330,60 @@ func (p *printer) flush() {
 		p.curComment++
 	}
 	p.output = append(p.output, '\n')
+// a VerbosePrinter prints a syntax tree and is intended for debugging purposes
```
ok

colincross commented on proptools/compare.go:
```diff
@@ -0,0 +1,266 @@
+	return false, explanation
+}
+
+// end of public members
```
This seems both unnecessary and misleading, since same() and different() are not public

mathjeff replied:
oops

colincross commented on proptools/compare.go:
```diff
@@ -0,0 +1,266 @@
+		}
+	}
+	if len(differences) > 0 {
+		return firstDifference(differences)
```
return false, sort.Strings(differences)[0]

mathjeff replied:
thanks

colincross commented on proptools/compare.go:
```diff
@@ -0,0 +1,266 @@
+	return same()
+}
+
+// compares two arrays (this ignores cap(), and only compares elements from indices 0 to len() )
```
arrays or slices

colincross commented on proptools/compare_test.go:
```diff
@@ -0,0 +1,229 @@
+		// a map with several differences, for confirming determinism
+		map[int]int{0: 0, 1: 1, 2: 2, 3: 3, 4: 4, 5: 5, 6: 6, 7: 7, 8: 8, 9: 9},
+		map[int]int{0: 1, 1: 2, 2: 3, 3: 4, 4: 5, 5: 6, 6: 7, 7: 8, 8: 9, 9: 0},
+		`a[0(%!t(int=0))] = 0 whereas b[0(%!t(int=0))] = 1`,
```
This seems like a bug?

mathjeff replied:
What would you prefer instead?

mathjeff commented on parser/printer.go:
```diff
@@ -15,351 +15,472 @@
-		p.requestNewlinesForPos(pos)
+func (p *printer) beforePrint(node ParseNode) {
+	var comments = p.syntaxTree.GetCommentsIfPresent(node)
+	if comments != nil {
```
Hmm, when I remove the check, I get a panic trying to access comments.preComments
